File: code/python/go2_stance.py
import sys
import math
from pathlib import Path

import numpy as np

BUILD_DIR = Path(__file__).resolve().parent.parent / "build"
sys.path.insert(0, str(BUILD_DIR))

import sqp_solver as sqp
import pinocchio as pin

RESOURCES = Path(__file__).resolve().parent.parent / "resources"

def main():
    N  = 100
    dt = 0.01

    urdf_path = str(RESOURCES / "urdf/go2/go2.urdf")
    print(f"URDF: {urdf_path}")

    ocp = sqp.OCP(N)

    feet = ["FL_foot", "FR_foot", "RL_foot", "RR_foot"]

    # Initial configuration
    model, collision_model, visual_model = pin.buildModelsFromUrdf(urdf_path, root_joint=pin.JointModelFreeFlyer())
    q0 = pin.neutral(model)
    q0[2] = 0.30
    q0[7:] = np.array([0., 0.8, -1.6, 0., 0.8, -1.6, 0., 0.8, -1.6, 0., 0.8, -1.6])


    data = pin.Data(model)
    pin.forwardKinematics(model, data, q0)
    pin.updateFramePlacements(model, data)

    # Find lowest foot z-coordinate
    min_foot_z = 1e10
    for foot in feet:
        fid = model.getFrameId(foot)
        fz = data.oMf[fid].translation[2]
        min_foot_z = min(min_foot_z, fz)
    # Adjust base height so feet are at ground level (z=0)
    q0[2] -= min_foot_z

    # Running nodes
    for k in range(N - 1):
        model, collision_model, visual_model = pin.buildModelsFromUrdf(urdf_path, root_joint=pin.JointModelFreeFlyer())
        node = sqp.Node(model)

        for foot in feet:
            node.add_contact(foot)

        node.add_dynamics(sqp.SemiEulerIntegration(dt))
        node.add_constraint(sqp.InvDynamics())

        for foot in feet:
            node.add_constraint(sqp.ContactConstraint(foot))
            node.add_constraint(sqp.FrictionConeConstraint(foot, 0.8))

        node.add_cost(sqp.ConfigurationCost(q0, 1e-3))
        node.add_cost(sqp.VelocityCost(1e-3))
        node.add_cost(sqp.AccelerationCost(1e-9))
        ocp.addNode(node)

    # Terminal node — high weight on reaching the target
    model, collision_model, visual_model = pin.buildModelsFromUrdf(urdf_path, root_joint=pin.JointModelFreeFlyer())
    node = sqp.Node(model)
    node.add_cost(sqp.ConfigurationCost(q0, 1e-3))
    node.add_cost(sqp.VelocityCost(1e3))
    ocp.addNode(node)

    ocp.finalize()

    # Initial guess: downward rest position, zero velocity and control
    print("q0:", q0)
    for k in range(N):
        ocp.get_node(k).q()[:] = q0
        ocp.get_node(k).v()[:] = 0.0
        ocp.get_node(k).u()[:] = 0.0


    # Solve
    solver = sqp.SQPSolver(ocp)
    opts = sqp.SQPoptions()
    opts.max_sqp_iters = 200
    opts.tolerance     = 1e-3
    #opts.ls_merit_eta  = 1e-4
    opts.ls_type       = sqp.LSType.MERIT
    solver.set_options(opts)

    solver.solve()

    ocp.save_trajectory("/workspace/code/resources/trajectories/go2_stance.json",
                        dt, urdf_path)


if __name__ == "__main__":
    main()

File: code/examples/go2/go2_stance.cpp
#include <sqp_solver/sqp_solver.h>
#include <trajopt/ocp.h>
#include <trajopt/node.h>
#include <trajopt/scheduler.h>
#include <trajopt/constraints/integration_schemes/semi-euler.h>
#include <trajopt/constraints/integration_schemes/euler.h>
#include <trajopt/constraints/inverse_dynamics.h>
#include <trajopt/constraints/joint_limits_constraint.h>
#include <trajopt/constraints/contact_constraint.h>
#include <trajopt/constraints/friction_cone_constraint.h>
#include <trajopt/costs/velocity_cost.h>
#include <trajopt/costs/acceleration_cost.h>
#include <trajopt/costs/configuration_cost.h>

#include <pinocchio/parsers/urdf.hpp>
#include <pinocchio/algorithm/joint-configuration.hpp>
#include <pinocchio/algorithm/kinematics.hpp>
#include <pinocchio/algorithm/frames.hpp>

#include <cmath>
#include <iostream>

int main() {
    // ── Load Go2 quadruped ──
    const std::string urdf_path = "/workspace/code/resources/urdf/go2/go2.urdf";
    pinocchio::Model model;
    pinocchio::urdf::buildModel(urdf_path, pinocchio::JointModelFreeFlyer(), model);

    std::cout << "Loaded " << model.name
              << " (nq=" << model.nq << ", nv=" << model.nv << ")\n";

    // ── Parameters ──
    int N = 5;
    const double dt = 0.01;
    const double mu = 0.8;  // friction coefficient

    // Foot frame names
    const std::vector<std::string> feet = {"FL_foot", "FR_foot", "RL_foot", "RR_foot"};


    std::cout << "Horizon: N=" << N << " nodes, T=" << N * dt << "s\n";

    // ── Nominal standing configuration ──
    // Floating base: [x, y, z, qx, qy, qz, qw] + 12 joint angles
    VectorXd q0 = pinocchio::neutral(model);
    q0[2] = 0.30;  // base height ~30cm

    // Hip, thigh, calf angles for a standing pose
    // FL
    q0[7]  =  0.0;    // FL_hip
    q0[8]  =  0.8;    // FL_thigh
    q0[9]  = -1.6;    // FL_calf
    // FR
    q0[10] =  0.0;    // FR_hip
    q0[11] =  0.8;    // FR_thigh
    q0[12] = -1.6;    // FR_calf
    // RL
    q0[13] =  0.0;    // RL_hip
    q0[14] =  0.8;    // RL_thigh
    q0[15] = -1.6;    // RL_calf
    // RR
    q0[16] =  0.0;    // RR_hip
    q0[17] =  0.8;    // RR_thigh
    q0[18] = -1.6;    // RR_calf

    // Compute actual base height from FK so feet touch ground
    {
        pinocchio::Data data(model);
        pinocchio::forwardKinematics(model, data, q0);
        pinocchio::updateFramePlacements(model, data);

        // Find lowest foot z-coordinate
        double min_foot_z = 1e10;
        for (const auto& foot : feet) {
            int fid = model.getFrameId(foot);
            double fz = data.oMf[fid].translation()[2];
            min_foot_z = std::min(min_foot_z, fz);
        }
        // Adjust base height so feet are at ground level (z=0)
        q0[2] -= min_foot_z;
    }

    // ── Build OCP ──
    OCP ocp(N);

    for (int k = 0; k < N; k++) {
        Node node(model);

        // Register all four feet as contacts
        node.add_contacts(feet);

        if (k < N - 1) {
            node.add_dynamics(std::make_shared<SemiEulerIntegration>(dt));
            node.add_constraint(std::make_shared<InvDynamics>());
            // node.add_constraint(std::make_shared<JointLimitsConstraint>());
        }

        // Contact and friction constraints for each foot
        for (const auto& foot : feet) {
            
            if (k < N - 1) {
                node.add_constraint(std::make_shared<ContactConstraint>(foot));
                node.add_constraint(std::make_shared<FrictionConeConstraint>(foot, mu));
            }
        }

        // Costs
        node.add_cost(std::make_shared<ConfigurationCost>(q0, 1e-3));
        if (k == N-1) node.add_cost(std::make_shared<VelocityCost>(1e3));
        if (k < N - 1) {
            node.add_cost(std::make_shared<VelocityCost>(1e-3));
            node.add_cost(std::make_shared<AccelerationCost>(1e-9));
        }

        ocp.addNode(std::move(node));
    }
    ocp.finalize();

    // ── Initial guess: standing pose with zero velocity ──
    std::cout<<"q0: "<<q0.transpose()<<std::endl;
    for (int k = 0; k < N; k++) {
        ocp.get_node(k).q() = q0;
        ocp.get_node(k).v().setZero();
        ocp.get_node(k).u().setZero();
    }

    // ── Solve ──
    SQPSolver solver(ocp);

    SQPoptions opts;
    opts.max_sqp_iters = 200;
    opts.ls_type = LSType::MERIT;
    opts.max_ls_iters = 5;
    opts.tolerance = 1e-3;
    // opts.verbose = 2;
    solver.set_options(opts);

    std::cout << "Solving trotting trajectory...\n";
    solver.solve();

    // ── Save trajectory ──
    ocp.save_trajectory("/workspace/code/resources/trajectories/go2_stance.json",
                        dt, urdf_path);

    return 0;
}
